File: Palindrome with Minimum Sum[11-May-2023]/Solution.java
//  >> JAVA CODE <<



class Solution {
    public static boolean possible(String s)
    {
        int low=0,high=s.length()-1;
        
        while(low<high)
        {
            if(s.charAt(low)!='?' && s.charAt(high)!='?' && s.charAt(low)!=s.charAt(high))
            {
                return false;
            }
            low++; high--;
        }
        
            return true;
    }
    public static int solve(String st)
    {
        char[] s=st.toCharArray();
        
        int low=0,high=s.length-1;
        
        while(low<high)
        {
            if(s[low]=='?')
                s[low]=s[high];
            else if(s[high]=='?')
                s[high]=s[low];
            
            low++; high--;
        }
        
        
        int i=0;
        while(i<s.length && s[i]=='?')
        {
            i++;
        }
        
        if(i==s.length) return 0;
        
        char prev=s[i];
        int ans=0;
        for(int x=i+1;x<s.length;x++)
        {
            if(s[x]=='?')
                s[x]=prev;
                
            ans+=Math.abs(s[x]-prev);
            prev=s[x];
        }
        
            return ans;
    }
    public static int minimumSum(String s) {
        return possible(s)?solve(s):-1;
    }
}

File: Palindrome with Minimum Sum[11-May-2023]/solution.py
#   >> PYTHON CODE << 




class Solution:
    def minimumSum(self, s : str) -> int:
        def possible(s):
            low,high=0,len(s)-1
            
            while(low<high):
                if(s[low] !='?' and s[high]!='?' and s[low]!=s[high]):
                    return False
                    
                low+=1
                high-=1
                
            return True
        
        def solve(st):
            s=list(st)
            i=0
            
            low,high=0,len(s)-1
            
            while(low<high):
                if(s[low]=='?'): s[low]=s[high]
                elif(s[high]=='?'): s[high]=s[low]
                low+=1
                high-=1
            
            #print(s)
            while(i<len(s) and s[i]=='?'): i+=1
            
            if(i==len(s)): return 0
            
            ans=0
            prev=s[i]
            for x in range(i,len(s)):
                if(s[x]=='?'):
                    s[x]=prev
                
                #print(abs(ord(s[x])-ord(prev)))
                ans+=abs(ord(s[x])-ord(prev))
                prev=s[x]
            
            return ans
            
        if not possible(s):
            return -1
        else:
            return solve(s)
     